### PR TITLE
Improved performance of the fp4tofp conversion

### LINKED ISSUE #4298: The performance of fp4_to_fp could be improved
### Describe the issue

The current implementations seems inefficient. For conversion of two fp4 values encoded in a single int8, it produces the following set of operations:

<details>

```llvm
    %32 = llvm.mlir.constant(7 : i8) : i8
    %33 = llvm.and %0, %32 : i8
    %34 = llvm.mlir.constant(112 : i8) : i8
    %35 = llvm.and %0, %34 : i8
    %36 = llvm.mlir.constant(12 : i16) : i16
    %37 = llvm.mlir.constant(8 : i8) : i8
    %38 = llvm.and %0, %37 : i8
    %39 = llvm.zext %38 : i8 to i16
    %40 = llvm.shl %39, %36 : i16
    %41 = llvm.mlir.constant(6 : i16) : i16
    %42 = llvm.zext %33 : i8 to i16
    %43 = llvm.shl %42, %41 : i16
    %44 = llvm.or %43, %40 : i16
    %45 = llvm.mlir.constant(8 : i16) : i16
    %46 = llvm.mlir.constant(-128 : i8) : i8
    %47 = llvm.and %0, %46 : i8
    %48 = llvm.zext %47 : i8 to i16
    %49 = llvm.shl %48, %45 : i16
    %50 = llvm.mlir.constant(2 : i16) : i16
    %51 = llvm.zext %35 : i8 to i16
    %52 = llvm.shl %51, %50 : i16
    %53 = llvm.or %52, %49 : i16
    %54 = llvm.mlir.constant(16128 : i16) : i16
    %55 = llvm.add %44, %54 : i16
    %56 = llvm.mlir.constant(0 : i8) : i8
    %57 = llvm.mlir.constant(6 : i8) : i8
    %58 = llvm.and %33, %57 : i8
    %59 = llvm.icmp "ne" %58, %56 : i8
    %60 = llvm.select %59, %55, %44 : i1, i16
    %61 = llvm.mlir.constant(16128 : i16) : i16
    %62 = llvm.add %53, %61 : i16
    %63 = llvm.mlir.constant(0 : i8) : i8
    %64 = llvm.mlir.constant(96 : i8) : i8
    %65 = llvm.and %35, %64 : i8
    %66 = llvm.icmp "ne" %65, %63 : i8
    %67 = llvm.select %66, %62, %53 : i1, i16
    %68 = llvm.mlir.constant(-32768 : i16) : i16
    %69 = llvm.and %60, %68 : i16
    %70 = llvm.mlir.constant(16128 : i16) : i16
    %71 = llvm.or %70, %69 : i16
    %72 = llvm.mlir.constant(1 : i8) : i8
    %73 = llvm.icmp "eq" %33, %72 : i8
    %74 = llvm.select %73, %71, %60 : i1, i16
    %75 = llvm.bitcast %74 : i16 to bf16
    %76 = llvm.mlir.constant(-32768 : i16) : i16
    %77 = llvm.and %67, %76 : i16
    %78 = llvm.mlir.constant(16128 : i16) : i16
    %79 = llvm.or %78, %77 : i16
    %80 = llvm.mlir.constant(16 : i8) : i8
    %81 = llvm.icmp "eq" %35, %80 : i8
    %82 = llvm.select %81, %79, %67 : i1, i16
    %83 = llvm.bitcast %82 : i16 to bf16
```

</details>

Instead of doing all these bit manipulations and conditionals, a single lookup table could be used.

### Environment details

main

### PR DESCRIPTION
**Pull request by AndreyPavlenko**
Use a simple lookup table instead of explicit conversion.
Use bitwise operations on vectors to build the indices.

Fixes #4298

------------
This is another (and perhaps overcomplicated) version of #4299, that uses bitwise operations on vectors (not per element) to build the indices. The resulting llir is the following:
```MLIR
%idx_vec0 = lshr <16 x i8> %i8vec, splat (i8 4)
%idx_vec1 = and <16 x i8> %i8vec, splat (i8 15)
%idx0 = extractelement <16 x i8> %idx_vec0, i64 0
%bf0 = extractelement <16 x bfloat> <bfloat 0xR0000, bfloat 0xR3F00, bfloat 0xR3F80, bfloat 0xR3FC0, bfloat 0xR4000, bfloat 0xR4040, bfloat 0xR4080, bfloat 0xR40C0, bfloat 0xR8000, bfloat 0xRBF00, bfloat 0xRBF80, bfloat 0xRBFC0, bfloat 0xRC000, bfloat 0xRC040, bfloat 0xRC080, bfloat 0xRC0C0>, i8 %idx0
%idx1 = extractelement <16 x i8> %idx_vec1, i64 0
%bf1 = extractelement <16 x bfloat> <bfloat 0xR0000, bfloat 0xR3F00, bfloat 0xR3F80, bfloat 0xR3FC0, bfloat 0xR4000, bfloat 0xR4040, bfloat 0xR4080, bfloat 0xR40C0, bfloat 0xR8000, bfloat 0xRBF00, bfloat 0xRBF80, bfloat 0xRBFC0, bfloat 0xRC000, bfloat 0xRC040, bfloat 0xRC080, bfloat 0xRC0C0>, i8 %idx1
...
 
```

### CODE REVIEW

**Comment by AndreyPavlenko** [comment]
A simple benchmark:

<details>

```python
import os
import timeit
import torch
import triton
import pathlib

def bench_fp4_to_bf16(sizePerThread = 16, inputSize = 8192, device=triton.runtime.driver.active.get_active_torch_device()):
    SPT = sizePerThread
    SPT2 = sizePerThread * 2
    ir = f"""
#blocked1 = #ttg.blocked<{{sizePerThread = [{SPT}], threadsPerWarp = [32], warpsPerCTA = [4], order = [0]}}>
#blocked2 = #ttg.blocked<{{sizePerThread = [{SPT2}], threadsPerWarp = [32], warpsPerCTA = [4], order = [0]}}>
module attributes {{triton_intel_gpu.support_bf16_conversion, triton_intel_gpu.support_dpas, triton_intel_gpu.support_sg_2d_block, ttg.global_scratch_memory_alignment = 1 : i32, ttg.global_scratch_memory_size = 0 : i32, triton_intel_gpu.target_arch = "spir16", "ttg.num-ctas" = 1 : i32, "ttg.num-warps" = 4 : i32, "ttg.threads-per-warp" = 32 : i32, ttg.shared = 16384 : i32, ttg.target = "xpu"}} {{

  tt.func public @fp4_to_bf16_kernel(%in_ptr: !tt.ptr<i8>, %out_ptr: !tt.ptr<bf16>) {{
    %c1 = arith.constant 1 : i64
    %s1 = arith.constant {SPT} : i32
    %s1_64 = arith.constant {SPT} : i64
    %s2 = arith.constant {SPT2} : i32
    %s2_16 = arith.constant {SPT2} : i64
    %pid = tt.get_program_id x : i32

    %in_off = arith.muli %pid, %s1 : i32
    %in_tensor_ptr = tt.make_tensor_ptr %in_ptr, [%s1_64], [%c1], [%in_off] {{order = array<i32: 0>, tt.divisibility = dense<[16]> : tensor<1xi32>}} : !tt.ptr<tensor<{SPT}xi8, #blocked1>>
    %in_tensor = tt.load %in_tensor_ptr {{alignment = 16 : i64}} : !tt.ptr<tensor<{SPT}xi8, #blocked1>>

    %bf16 = ttg.fp4_to_fp %in_tensor {{axis = 0 : i32}} : tensor<{SPT}xi8, #blocked1> -> tensor<{SPT2}xbf16, #blocked2>

    %out_off = arith.muli %pid, %s2 : i32
    %out_tensor_ptr = tt.make_tensor_ptr %out_ptr, [%s2_16], [%c1], [%out_off] {{order = array<i32: 0>, tt.divisibility = dense<[16]> : tensor<1xi32>}} : !tt.ptr<tensor<{SPT2}xbf16, #blocked2>>
    tt.store %out_tensor_ptr, %bf16 {{alignment = 16 : i64}} : !tt.ptr<tensor<{SPT2}xbf16, #blocked2>>
    tt.return
  }}
}}
"""
    tmp_path: pathlib.Path = pathlib.Path(".")
    temp_file = tmp_path / "fp4_to_bf16_kernel.ttgir"
    temp_file.write_text(ir)
    kernel = triton.compile(str(temp_file))
    os.remove(temp_file)

    
    x = torch.randint(0, 127, (inputSize,), dtype=torch.int8, device=device)
    y = torch.zeros((inputSize * 2,), dtype=torch.bfloat16, device=device)

    def run_kernel():
        kernel[(inputSize // sizePerThread, 1, 1)](x, y)
        torch.xpu.synchronize(device)
    
    run_kernel()
    # print(x)
    # print(y)

    time = timeit.timeit(run_kernel, number=100000)
    print(f"Time: {time}. Size per thread: {sizePerThread}, Input size: {inputSize}")


if __name__ == "__main__":
    bench_fp4_to_bf16(16)
    bench_fp4_to_bf16(32)
    bench_fp4_to_bf16(64)
    bench_fp4_to_bf16(128)
    bench_fp4_to_bf16(256)
    bench_fp4_to_bf16(512)
```

</details>

The results:
```
Time: 2.9286755439825356. Size per thread: 16, Input size: 8192
Time: 2.9111483457963914. Size per thread: 32, Input size: 8192
Time: 2.93849644693546. Size per thread: 64, Input size: 8192
Time: 2.922249269904569. Size per thread: 128, Input size: 8192
Time: 3.4274095639120787. Size per thread: 256, Input size: 8192
Time: 4.842097139917314. Size per thread: 512, Input size: 8192
```
Main branch results:
```
Time: 3.15685032797046. Size per thread: 16, Input size: 8192
Time: 3.149760977132246. Size per thread: 32, Input size: 8192
Time: 3.442877721041441. Size per thread: 64, Input size: 8192
Time: 3.5604478721506894. Size per thread: 128, Input size: 8192
Time: 4.6088251559995115. Size per thread: 256, Input size: 8192
Time: 6.909060714067891. Size per thread: 512, Input size: 8192
```

**Comment by AndreyPavlenko** [comment]
In the last commit i32vec to i8vec conversion is elementwise, instead of bitcasting the entire i32vec to i8vec, but it seems slightly faster.

```
Time: 2.877734236884862. Size per thread: 16, Input size: 8192
Time: 2.801578857935965. Size per thread: 32, Input size: 8192
Time: 2.9648928409442306. Size per thread: 64, Input size: 8192
Time: 2.8129132031463087. Size per thread: 128, Input size: 8192
Time: 3.0528822529595345. Size per thread: 256, Input size: 8192
Time: 4.295081242918968. Size per thread: 512, Input size: 8192
```
In the asm I see 2 `and`, but 4 `shr` instructions for each block of 16 bytes:
```asm
// Line 16
(W)     load.ugm.d32x4t.a64 (1|M0)  r5:1        [r6:1]             {I@2,$3} // ex_desc:0x0; desc:0x210B580 // $19
...
// Line 18
(W)     and (4|M0)               r6.8<1>:d     r5.0<1;1,0>:d     252645135:d               {$3.dst}  //  ALU pipe: int; $61
(W)     shr (1|M0)               r6.4<1>:ud    r5.0<0;1,0>:ud    4:w                                 //  ALU pipe: int; $21
(W)     shr (1|M0)               r6.5<1>:ud    r5.1<0;1,0>:ud    4:w                                 //  ALU pipe: int; $22
(W)     shr (1|M0)               r6.6<1>:ud    r5.2<0;1,0>:ud    4:w                                 //  ALU pipe: int; $23
(W)     shr (1|M0)               r6.7<1>:ud    r5.3<0;1,0>:ud    4:w                                 //  ALU pipe: int; $24
(W)     and (4|M0)               r5.12<1>:d    r6.4<1;1,0>:d     252645135:d               {I@3}     //  ALU pipe: int; $25
```
The corresponding llir:
```MLIR
  %17 = load <4 x i32>, ptr addrspace(1) %16, align 16, !dbg !20
  %18 = lshr <4 x i32> %17, splat (i32 4), !dbg !21
  %19 = and <4 x i32> %18, splat (i32 252645135), !dbg !21
  %30 = and <4 x i32> %17, splat (i32 252645135), !dbg !21
```
252645135 is the 0x0F0F0F0F mask.

**Review by LiyangLingIntel — DISMISSED** [review]
LGTM. 
How about adding several lit tests to validate this change since the e2e UT has still been blocked?

**Comment by AndreyPavlenko** [comment]
> How about adding several lit tests to validate this change since the e2e UT has still been blocked?

Done.

**Comment by etiotto** [comment]
The implementation has changed since approval. @LiyangLingIntel and @ienkovich can you please take another look?

File: third_party/intel/lib/TritonIntelGPUToLLVM/Fp4ToFpOpToLLVM.cpp [hunk 1]
```diff
@@ -13,6 +13,38 @@ using namespace mlir::triton::gpu;
 using namespace mlir::triton::gpu::intel;
 
 namespace {
+struct CachingBuilder : TritonLLVMOpBuilder {
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/Fp4ToFpOpToLLVM.cpp]: what's the motivation? I would expect CSE should be able to easily fold constants?

Comment (AndreyPavlenko) [reply third_party/intel/lib/TritonIntelGPUToLLVM/Fp4ToFpOpToLLVM.cpp]: The motivation is to make the intermediate IRs a bit more readable. Some [tests](https://github.com/intel/intel-xpu-backend-for-triton/issues/4062) produce gigabytes of IRs and it's not easy to analyse them. In terms of performance it could also be beneficial. Thousands of duplicate constants add more job for CSE, but I've not compared the timings. As for me, I'd probably make a global cache, that can be reused by other passes.

Comment (etiotto) [reply third_party/intel/lib/TritonIntelGPUToLLVM/Fp4ToFpOpToLLVM.cpp]: we have an utility named `indOrCreateIntConstant` which doesn't use a cache but inspect the IR to determine whether a constant is already available in the function and can be reused. See intel/libs/Utils/Utility.cpp

Comment (whitneywhtsang) [reply third_party/intel/lib/TritonIntelGPUToLLVM/Fp4ToFpOpToLLVM.cpp]: icic, a global cache that's independent of fp4tofp seems like a better idea. maybe one can change the existing `i8_val`, `i32_val`, `int_val` defined in `include/triton/Conversion/TritonGPUToLLVM/Utility.h`.

Comment (AndreyPavlenko) [reply third_party/intel/lib/TritonIntelGPUToLLVM/Fp4ToFpOpToLLVM.cpp]: Do you mean changing the implementation of TritonLLVMOpBuilder?

Comment (whitneywhtsang) [reply third_party/intel/lib/TritonIntelGPUToLLVM/Fp4ToFpOpToLLVM.cpp]: yes, and upstreaming the change.

Comment (AndreyPavlenko) [reply third_party/intel/lib/TritonIntelGPUToLLVM/Fp4ToFpOpToLLVM.cpp]: I could try, but not sure if this change will be accepted upstream. It will also require some logic to check the constants visibility. If a constant is created and cached in one block, it's inaccessible in another and needs to be either duplicated or moved to the common parent.

Comment (whitneywhtsang) [reply third_party/intel/lib/TritonIntelGPUToLLVM/Fp4ToFpOpToLLVM.cpp]: To start, we can simplify it to caching per basic block, then we don't need to worry about dominator tree.

Comment (ienkovich) [reply third_party/intel/lib/TritonIntelGPUToLLVM/Fp4ToFpOpToLLVM.cpp]: You can always insert constants at the beginning of the function to dominate every position. Likely to break many LIT tests if applied globally, though.

Comment (AndreyPavlenko) [reply third_party/intel/lib/TritonIntelGPUToLLVM/Fp4ToFpOpToLLVM.cpp]: Right, this is another solution. It would also be good if all the builders could use the same cache. Do you have an idea how to share the cache?

Comment (AndreyPavlenko) [reply third_party/intel/lib/TritonIntelGPUToLLVM/Fp4ToFpOpToLLVM.cpp]: I tried to make a [shared constants cache](https://github.com/intel/intel-xpu-backend-for-triton/commit/8aff1371ee773400a9217fe1de7a78dafa4da99f#diff-ea09e37788131c30733ddee48bf7871b744be17b233c5650f679e3e2b8f7c7ee) and it seems to be working, the tests pass. This implementation creates constants in the beginning of each isolated block and saves in the cache, that is shared among the builders. As expected, a lot of lit tests fail. I've fixed some of them, but still ~20 need to be fixed.

I've also added 2 [pref tests](https://github.com/intel/intel-xpu-backend-for-triton/commit/8aff1371ee773400a9217fe1de7a78dafa4da99f#diff-11f44b56f55871122637102ebb3e5534c2f5e489278ba2c48de95aa67a598f27R111), that measure times of constants creation and the canonicalizer pass. The first one creates a lot of unused constants, that are removed by canonicalizer. The second one is a bit more realistic - it not only creates the constants, but inserts them into vectors.
Here are the results:
```
Without cache:
Create| Insertion time: 0.378667
Create| Canonicalizer time: 0.341482
Fill| Insertion time: 0.648058
Fill| Canonicalizer time: 1.23455

With cache:
Create| Insertion time: 0.178808
Create| Canonicalizer time: 0.000576043
Fill| Insertion time: 0.385093
Fill| Canonicalizer time: 0.639965
```
What do you think, could this feature be useful and should I upstream it?

Comment (AndreyPavlenko) [reply third_party/intel/lib/TritonIntelGPUToLLVM/Fp4ToFpOpToLLVM.cpp]: > we have an utility named `indOrCreateIntConstant` which doesn't use a cache but inspect the IR

I tried a similar approach and it was terribly slow. Tried it once again with [this test](https://github.com/intel/intel-xpu-backend-for-triton/pull/4440#discussion_r2183797644). Here are the results:
```
Create| Insertion time: 8.21444
Create| Canonicalizer time: 0.00148386
Fill| Insertion time: 7.21295
Fill| Canonicalizer time: 0.761791
```
The above results are for the release build, for the debug it's much slower and CI fails by timeout:
```
Create| Insertion time: 137.241
Create| Canonicalizer time: 0.00160167
Fill| Insertion time: 216.471
Fill| Canonicalizer time: 2.18726
```
